### DISCUSSION TITLE
DataNotFoundException 선언

### DISCUSSION THREAD

**Discussion by Choi-JJunho** [BCSDLab/KOIN_API_V2#163] [에러 / 고민사항]
현재 UserNotFoundException, ArticleNotFoundException 등 `데이터를 찾을 수 없다` 라는 상황에 해당하는 예외들이 구체적으로 구현되어있습니다.
에러를 너무 잘게 쪼개는게 아닌지 고민이 듭니다.
특정 데이터를 찾을 수 없는 경우 DataNotFoundException을 예외로 발생시키도록 구성하는건 어떨까요?

특정 데이터를 찾을 수 없다는 예외에 대해 클라이언트에게 404(NotFound)를 반환한다는 것 또한  DataNotFoundException으로 묶어서 표현하면 좋겠다는 근거 중 하나입니다.

**Comment by Choi-JJunho** [comment]
문득 추가로 든 생각인데 UserNotFoundException, ArticleNotFoundException등 구체화된 예외들이 DataNotFoundException을 상속받는 구조도 괜찮아보이기도 하네요

**Reply by songsunkook** [reply]
저도 구체화된 예외가 DataNotFoundException을 상속받는 구조가 좋아 보입니다! 예외별로 구체적인 처리가 필요한 경우에는 같은 종류라도 예외가 서로 분리되어 핸들링되는 것이 적절할 것 같습니다.

**Comment by Invidam** [comment]
- 404가 꼭 DataNotFound만 있진 않을 것 같아요 (페이지 못찾았을 때 등등)
- 예외 구체화를 통해서 얻을 수 있는 이점은 ExceptionHandler에서 다른 처리를 할 수 있다는 점 같아서 서로 같은 처리를 한다면 구체화하지 않는 것(여러 객체로 세분화하지 않는 것)이 복잡성이 줄어든다는 이점이 있을 것 같아요

기타)
```
package in.koreatech.koin.global.exception;

public class Error {
    public static class NotFound {

    }

    public static class Unauthorized {

    }
}

```
이런식으로 예외들을 모아두면 안되냐는 의견도 있었어요
